lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsInput.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.gcs;
+
+import com.google.cloud.ReadChannel;
+import com.google.cloud.storage.Blob;
+import io.trino.filesystem.TrinoInput;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.OptionalLong;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.trino.filesystem.gcs.GcsUtils.getReadChannel;
+import static io.trino.filesystem.gcs.GcsUtils.handleGcsException;
+import static java.util.Objects.checkFromIndexSize;
+import static java.util.Objects.requireNonNull;
+
+final class GcsInput
+        implements TrinoInput
+{
+    private final GcsLocation location;
+    private final Blob blob;
+    private final int readBlockSize;
+    private OptionalLong length;
+    private boolean closed;
+
+    // Note: default read chunk size is 2Mib - see com.google.cloud.storage.BaseStorageReadChannel
+    public GcsInput(GcsLocation location, Blob blob, int readBlockSize, OptionalLong length)
```

Comment:
We shouldn't need to pass `Blob` here, as we should be able to do reads without needing to fetch the object metadata first.